hazelcast/src/hazelcast/cp/cp.cpp

```diff
@@ -653,6 +665,257 @@ namespace hazelcast {
+                            } catch (exception::SessionExpiredException &) {
+                                session_manager_.invalidate_session(group_id_, session_id);
+                                if (use_timeout && (timeout - (std::chrono::steady_clock::now() - start) <= std::chrono::milliseconds::zero())) {
+                                    return std::make_pair(false, false);
+                                }
+                                return std::make_pair(false, true);
+                            } catch (exception::WaitKeyCancelledException &) {
+                                session_manager_.release_session(group_id_, session_id, permits);
+                                if (!use_timeout) {
+                                    BOOST_THROW_EXCEPTION(
+                                            exception::IllegalStateException("session_semaphore::try_acquire_for_millis", (boost::format(
+                                                    "Semaphore[%1%] not acquired because the acquire call on the CP group is cancelled, possibly because of another indeterminate call from the same thread.") %
+                                                                                                            object_name_).str()));
+                                }
+                                return std::make_pair(false, false);
+                            }
+                        });
+            });
+
+            return do_try_acquire_once().then(boost::launch::deferred, [=] (boost::future<std::pair<bool, bool>> f) {
+                auto result = f.get();
+                if (result.second) {
+                    return result.first;
+                }
+                for (;result.second;result = do_try_acquire_once().get());
+                return result.first;
+            });
+        }
+
+        boost::future<void> session_semaphore::release(int32_t permits) {
+            auto session_id = session_manager_.get_session(group_id_);
+            if (session_id == internal::session::proxy_session_manager::NO_SESSION_ID) {
+                BOOST_THROW_EXCEPTION(illegal_state_exception(nullptr));
+            }
+
+            auto thread_id = get_thread_id();
+            return do_release(permits, thread_id).then([=] (boost::future<void> f) {
+                try {
+                    f.get();
+                    session_manager_.release_session(group_id_, session_id, permits);
+                } catch(exception::SessionExpiredException &) {
+                    session_manager_.invalidate_session(group_id_, session_id);
+                    session_manager_.release_session(group_id_, session_id, permits);
+                    BOOST_THROW_EXCEPTION(illegal_state_exception(std::current_exception()));
+                }
+            });
+        }
+
+        exception::IllegalStateException session_semaphore::illegal_state_exception(std::exception_ptr e) {
```

Comment:
Why this function have a return type? It is only throwing, not returning.
It could be `[[noreturn]] void session_semapho...`

Reply:
changed it to `throw_illegal_state_exception` with void return.